OPHD/Technology/TechnologyCatalog.cpp

```diff
@@ -36,6 +37,71 @@ namespace
 		{"structure", Technology::Unlock::Unlocks::Structure},
 		{"vehicle", Technology::Unlock::Unlocks::Vehicle}
 	};
+
+
+	void readEffects(NAS2D::Xml::XmlElement& effects, Technology& technology)

```

Comment:
You can probably do similar with this method too. Instead of taking a `Technology&` as a parameter, and then modifying it, you could instead return the parsed effects arrays.

Here the code is complicated by having two arrays. I suppose you could return a `struct`/`tuple` of the two arrays, though you could also split the function into two, where each function only processes are particular element type. I believe the `for` loop iteration can use a parameter to `firstChildElement` and `nextSiblingElement` so you only process elements with a given name. That should make it relatively easy to split the function into two, each one returning a `std::vector` for that element type.

Reply:
I was thinking about that and how to do it. Initially I was going to have it return an effect but since they are no longer derived I need to figure out a different way to do it. It shouldn't be difficult, just need to come up with a clean way to do it that doesn't also pollute `readTechnology`.

Reply:
For reference, maybe look at the code for:
```cpp
for (const auto* node = element->firstChildElement("imagesheet"); node; node = node->nextSiblingElement("imagesheet"))
```

Or:
```cpp
for (auto* mineElement = element->firstChildElement("mines")->firstChildElement("mine"); mineElement; mineElement = mineElement->nextSiblingElement())
```

There are a few places in the code that use a parameter to restrict iteration to only `XmlElement`s of a certain type.